Update Passenger Services Mode for new commuter shuttle fields

matthewmarwedel opened the pull request:
---
name: Matthew Marwedel, SFMTA @matthewmarwedel supported by @Mu-yi-Zhou, @monachiu, and @alexdemisch (SFMTA)
about: Suggest changes to MDS to support commuter shuttles within passenger services mode
title: Adding new fields to Passenger Services mode to support Commuter Shuttles

---

## Explain pull request

SFMTA would like to update the passenger services mode in MDS to include support for commuter shuttles. Most fields in passenger services match the needs of commuter shuttle (such as stop_id, lat, long, etc.) We would like to add three additional fields for commuter shuttle use. These fields would be optional (required for commuter shuttle use case, but optional as an MDS field). When those fields are implemented, we would like to add support to integrate [GTFS](https://gtfs.org/documentation/schedule/reference/) endpoints to passenger services mode to track shuttle routes. 

Initial pull request includes:
adding **service_id** field to **[Events, Trips, Telemetry]**
adding **route_id** field to **[Events, Trips, Telemetry]**
adding **stop_sequence** field to **[Events, Trips, Telemetry]**

Once this is implemented we will build on the pull request for the following:
adding GTFS integration in MDS for **route_id**  and **service_id**

## Is this a breaking change

A breaking change would require consumers or implementors of the API to modify their code for it to continue to function (ex: renaming of a required field or the change in data type of an existing field). A non-breaking change would allow existing code to continue to function (ex: addition of an optional field or the creation of a new optional endpoint).

* No, not breaking

We would like to add some optional fields fields to passenger services mode along with an optional **Schedule** endpoint in the future to support commuter shuttles. Existing code will continue to function.

## Impacted Spec

Which spec(s) will this pull request impact?

* `agency`
* `provider`
* `policy`

## Additional context

N/A

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/openmobilityfoundation/mobility-data-specification?pullRequest=927) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/openmobilityfoundation/mobility-data-specification?pullRequest=927) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/openmobilityfoundation/mobility-data-specification?pullRequest=927) it.</sub>

schnuerle commented:
Some items to consider.

1. Should these fields be in the data types for all modes, or in just Passenger Service
2. Potentially add a `service_id` to better describe the route id
3. Could add another field `route_name`, or reference it only in external endpoint in Schedule
4. Maybe add `stop_sequence` to telemetry too
5. Policy and Geography may not need route and stop info, geographic policies should be sufficient
6. A draft of what's in Schedule would be useful to see, like route info, stop sequence, calendar (days of week), calendar dates (month/day of operation), including exceptions 
   - GTFS could be the solution here. Would need to make sure `route_id`, `service_id` is in GTFS Schedule format, and new GTFS URL ([info](https://gtfs.org/documentation/schedule/reference/)) field in MDS Trips. `GTFS_Stop_id` may be needed in events too. 
8. stop sequence might not need to go in Trips
9. Might need a `gtfs_url` in the providers.csv file to track new operator IDs
10. Operators would generate a unique `device_id` UUID, but can also use VIN, license plate, USDOT id, etc fields.

schnuerle commented:
We will be discussing this at this week's [MDS Working Group meeting](https://github.com/openmobilityfoundation/mobility-data-specification/wiki/Web-conference-notes,-2025.02.13-(MDS-Working-Group)). Please review and leave your comments and use cases and questions ahead of time.

matthewmarwedel commented:
Hi all,

Replying to earlier discussion points with my thoughts

1. Should these fields be in the data types for all modes, or in just Passenger Service
**I think there is a lot of potential for other modes and the new fields should be for all modes. I would love to hear thoughts from other OMF members**
2. Potentially add a `service_id` to better describe the `route id`
**`service_id` and `route id` will _both_ be needed to show what the route is and what times it runs. These can reference routes.txt, calendar.txt, and calendar_dates.txt in [GTFS](https://gtfs.org/documentation/schedule/reference/).**
3. Could add another field `route_name`, or reference it only in external endpoint in Schedule
**Referencing routes.txt in [GTFS](https://gtfs.org/documentation/schedule/reference/) may be the best option here, no `route_name` field needed**
4. Maybe add `stop_sequence` to telemetry too
**Yes, add `stop_sequence` to Telemetry**
5. Policy and Geography may not need route and stop info, geographic policies should be sufficient
**Policy and Geography endpoints don't need the new fields (`stop_sequence`, `service_id` and `route id`)**
6. A draft of what's in Schedule would be useful to see, like route info, stop sequence, calendar (days of week), calendar dates (month/day of operation), including exceptions

- GTFS could be the solution here. Would need to make sure `route_id`, `service_id` is in GTFS Schedule format, and new GTFS URL ([info](https://gtfs.org/documentation/schedule/reference/)) field in MDS Trips. GTFS_Stop_id may be needed in events too.
**Use routes.txt, calendar.txt, and calendar_dates.txt in [GTFS](https://gtfs.org/documentation/schedule/reference/) as the schedule with `route_id` and `service_id` matching GTFS format. I'm not sure if that alone is enough to connect MDS to the GTFS spec, if not a `gtfs_url` field may be needed in [Trips, Events, Telemetry]. I'm also not sure if the `gtfs_url` field would be needed in all of them. Would love insight from OMF community on this.**
7. `stop sequence` might not need to go in Trips
**Not sure, depends if stops are in MDS or GTFS. For now, it seems like `stop sequence`  will stay in MDS and STFS stops won't be needed.**
8. Might need a gtfs_url in the providers.csv file to track new operator IDs
**Yes, add a `gtfs_api_url` field to the MDS [provider list](https://github.com/openmobilityfoundation/mobility-data-specification/blob/dev/providers.csv)**
9. Operators would generate a unique device_id UUID, but can also use VIN, license plate, USDOT id, etc fields.
**This sounds good.**

matthewmarwedel commented:
I added a new commit to the pull request finalizing a few changes discussed above.

- `stop_sequence` was added to **Telemetry**
- `service_id` was added to **Events, Trips, Telemetry**
- GTFS integration support was added to documentation in **Events, Trips, Telemetry** 
- `gtfs_api_url` was added to providers.csv

schnuerle commented on data-types.md:
```diff
@@ -140,6 +143,9 @@ A standard point of vehicle telemetry. References to latitude and longitude impl
 | `battery_percent` | Integer         | Required if Applicable | Percent battery charge of vehicle, expressed between 0 and 100 |
 | `fuel_percent`    | Integer         | Required if Applicable | Percent fuel in vehicle, expressed between 0 and 100 |
 | `tipped_over`     | Boolean         | Required if Known      | If detectable and known, is the device tipped over or not? Default is 'false'. |
+| `stop_sequence` | Integer | Required if Applicable | Order of stops for a particular trip. The values must increase along the trip, but do not need to be consecutive. |
```
Just making a note here for the 'Files changed' view that these fields are for **Telemetry**.

schnuerle commented on data-types.md:
```diff
@@ -237,6 +243,9 @@ A Trip is defined by the following structure:
 | `standard_cost`          | Integer         | Optional | The cost, in the currency defined in `currency`, to perform that trip in the standard operation of the System (see [Costs & Currencies][costs-and-currencies]) |
 | `actual_cost`            | Integer         | Optional | The actual cost, in the currency defined in `currency`, paid by the customer of the *mobility as a service* provider (see [Costs & Currencies][costs-and-currencies]) |
 | `currency`               | String          | Optional, USD cents is implied if null.| An [ISO 4217 Alphabetic Currency Code][iso4217] representing the currency of the payee (see [Costs & Currencies][costs-and-currencies]) |
```
Just making a note here for the 'Files changed' view that these fields are for **Trips**.

schnuerle commented on providers.csv:
```diff
@@ -1,75 +1,75 @@
-Swoop,passenger-services,08ad2b21-8115-4d18-874f-48082e8e50fb,https://www.swoopapp.com/,,
-GoSharing,micromobility,1f859479-4427-4f52-8220-05c786436314,https://www.binbin.tech/,https://complianceapi.binbinapp.com/api/mds/,https://complianceapi.binbinapp.com/api/gbfs/
\ No newline at end of file
+provider_name,mode_id,provider_id,url,mds_api_url,gbfs_api_url,gtfs_api_url
```
There should be a note added somewhere that this URL is to the ZIP file of the GTFS Schedule feed files.

schnuerle replied:
Not using this field now

schnuerle commented on data-types.md:
```diff
@@ -114,6 +114,9 @@ Events represent changes in vehicle status.
 | `associated_ticket` | String | Optional | Identifier for an associated ticket inside an Agency-maintained 311 or CRM system |
+| `stop_sequence` | Integer | Required if Applicable | Order of stops for a particular trip. The values must increase along the trip, but do not need to be consecutive. |
+| `route_id` | UUID | Required if Applicable | A unique route ID for the associated route and stops. Matches [GTFS](https://gtfs.org/documentation/schedule/reference/) `route_id` |
+| `service_id` | UUID | Required if Applicable | A unique service ID for the set of dates when service is available for a route. Matches [GTFS](https://gtfs.org/documentation/schedule/reference/) `service_id` |
```
Just making a note here for the 'Files changed' view that these fields are for **Events**.

schnuerle commented on data-types.md:
```diff
@@ -140,6 +143,9 @@ A standard point of vehicle telemetry. References to latitude and longitude impl
 | `tipped_over`     | Boolean         | Required if Known      | If detectable and known, is the device tipped over or not? Default is 'false'. |
+| `stop_sequence` | Integer | Required if Applicable | Order of stops for a particular trip. The values must increase along the trip, but do not need to be consecutive. |
+| `route_id` | UUID | Required if Applicable | A unique route ID for the associated route and stops. Matches [GTFS](https://gtfs.org/documentation/schedule/reference/) `route_id` |
+| `service_id` | UUID | Required if Applicable | A unique service ID for the set of dates when service is available for a route. Matches [GTFS](https://gtfs.org/documentation/schedule/reference/) `service_id` |
```
Just making a note here for the 'Files changed' view that these fields are for **Telemetry**.

schnuerle commented on data-types.md:
```diff
@@ -237,6 +243,9 @@ A Trip is defined by the following structure:
 | `currency`               | String          | Optional, USD cents is implied if null.| An [ISO 4217 Alphabetic Currency Code][iso4217] representing the currency of the payee (see [Costs & Currencies][costs-and-currencies]) |
+| `stop_sequence` | Integer | Required if Applicable | Order of stops for a particular trip. The values must increase along the trip, but do not need to be consecutive. |
+| `route_id` | UUID | Required if Applicable | A unique route ID for the associated route and stops. Matches [GTFS](https://gtfs.org/documentation/schedule/reference/) `route_id` |
+| `service_id` | UUID | Required if Applicable | A unique service ID for the set of dates when service is available for a route. Matches [GTFS](https://gtfs.org/documentation/schedule/reference/) `service_id` |
```
Just making a note here for the 'Files changed' view that these fields are for **Trips**.

schnuerle commented:
Open questions:

- Should `stop_sequence`, `route_id`, `service_id` be in the base MDS objects so they can be used for more than one mode (eg, maybe for Delivery Robots too), or should they only be in the Passenger Services mode?
- Do `stop_sequence`, `route_id`, `service_id` need to all be in Events, Telemetry, and Trips?

matthewmarwedel commented:
Hi all,

After the OMF discussion, I will be reviewing the following updates to the pull request with the SFMTA team:

- removing `stop_sequence` from **Trips, Telemetry**
- adding `stop_id_gtfs` to **Events**
- adding supporting documentation for the `gtfs_api_url` field in providers.csv
- adding a `gtfs_api_url` field to **Trips, Events** for linking `route_id`, `service_id`, and `stop_id_gtfs` to the correct gtfs url

schnuerle commented:
> * adding `stop_id_gtfs` to **Events**

I think call this `gtfs_stop_id`, per some other MDS/CDS naming conventions. 

> * adding supporting documentation for the `gtfs_api_url` field in providers.csv (the Adding an MDS Provider ID page)

The `gtfs_api_url` field can be removed from providers.csv, thus reverting it back to the way it was. Instead adding it to Trips, Events would be enough. Operators can still add their name, website, UUID, and applicable mode to the file, but it doesn't make sense to add a single URL for GTFS.

matthewmarwedel commented:
Summary of changes I made from suggestions:
- removed `stop_sequence` from **Trips, Telemetry**
- added `gtfs_stop_id` to **Events**
- added `shape_id` to **Trips**
- removed `gtfs_api_url` column in providers.csv
- added a `gtfs_api_url` field to **Trips, Events** for linking `route_id`, `shape_id`, `service_id`, and `gtfs_stop_id` to the correct gtfs url

stevenmwhite commented:
I have not been involved in the development of MDS over the past few years, but back in 2020 I was collaborating with Ellis & Associates / Lacuna Tech. on a proposal to add fixed-route service to MDS for LADOT (my company manages the fixed route dispatch software and GTFS/GTFS-RT for LADOT). It never really got off the ground, but I'm going to share the notes of where we had progressed to.

I'm not proposing or advocating for anything here, simply sharing for your information. Do keep in mind that this was based on the state of the spec in 2020 and may include assumptions that are well out of date today.
[mds_fixed_route.md](https://github.com/user-attachments/files/19059878/mds_fixed_route.md)

matthewmarwedel commented:
Summary of changes I made after discussion with SFMTA team on matching MDS to GTFS format:

- Clarified definitions of `route_id`, `service_id` in **Trips, Telemetry, Events** to match GTFS route-trip schedule format
- Added `gtfs_trip_id` to **Trips, Telemetry, Events**
- Removed `gtfs_trip_id` from **Events**; MDS existing `stop_id` field will be used instead

schnuerle commented:
1. Shouldn’t the `gtfs_shape_id` in MDS Trips for the [shapes.txt](https://gtfs.org/documentation/schedule/reference/#shapestxt) file be discoverable via the associated route_id, service_id, and trip_id in the GTFS trips.txt file? If so the `gtfs_shape_id' field in MDS Trips is not needed.

2. For the MDS Events `stop_sequence` field, are these be the same as the GTFS `stop_sequence` field in stop_times.txt, and should we reference that in the description with a link? 

   - Is stop sequence even needed here, and instead use the `stop_id` field from GTFS (which includes the expected sequence). Even if there is an unexpected stop, the stop sequence can be determined by timestamps in MDS events or telemetry.

5. MDS Events has a `trip_ids` field to connect to MDS Trips. Can gtfs_route_id, gtfs_service_id, and gtfs_trip_id just be cross referenced via the trip id in MDS Trips?  Even if a bus is not on a trip, then referencing GTFS trip info won’t be needed or possible anyway.

6. MDS Telemetry has a `trip_ids` field to connect to MDS Trips, so can  gtfs_route_id, gtfs_service_id, and gtfs_trip_id just be cross referenced via the trip id? Similar to item 3 above.

7. Do MDS Events and MDS Telemetry need a `gtfs_api_url` field, if they happen when they are not on a GTFS trip? If on a GTFS trip, can it just reference this `gtfs_api_url` field in MDS Trips via `trip_id`?

8. MDS Telemetry and MDS Events each have their own timestamp and location fields built in. MDS Telemetry has an MDS `stop_id` field to reference MDS Stops, and Events has a new GTFS `stop_sequence` field. But shouldn’t Telemetry and Events have a stops.txt GTFS `gtfs_stop_id` field to know when the bus stopped at a scheduled GTFS stop along the route?

9. What is the need of`gtfs_service_id` in MDS Trips? `service_id` in GTFS connects to the calendar.txt file, which only are used to "define exceptions to the default service patterns." I may be missing a use case here.
 
10. In MDS Trips, `gtfs_route_id` references a `route_id` in GTFS, but this ID is already part of the trips.txt field that is referenceable via the `gtfs_trip_id` field. So it may not be needed to be duplicated here.

Let me see if I can summarize all of this. 

**MDS Events**

Current:

- `stop_sequence` 
- `gtfs_route_id`
- `gtfs_service_id`
- `gtfs_trip_id`
- `gtfs_api_url`

_Proposed:_

- `gtfs_stop_id` 
- use `trip_ids` to make GTFS field connections

**MDS Telemetry**

Current:

- `gtfs_route_id`
- `gtfs_service_id`
- `gtfs_trip_id`

_Proposed:_

- `gtfs_stop_id`
- use `trip_ids` to make GTFS field connections

**MDS Trips**

Current:

- `gtfs_route_id`
- `gtfs_service_id`
- `gtfs_shape_id` 
- `gtfs_trip_id`
- `gtfs_api_url`

_Proposed:_

- `gtfs_trip_id`
- `gtfs_api_url`

I may have to turn the proposed fields into a diagram to show the connections between MDS objects and GTFS too.

matthewmarwedel commented:
Hi @schnuerle,

We reviewed within SFMTA and agreed with most of your proposal. I will update the pull request with the below.

**MDS Events**
- remove `gtfs_route_id`, `gtfs_trip_id`, `gtfs_service_id`, `gtfs_api_url`; These can all be found by linking to MDS trips with MDS trip_id.
- remove `stop_sequence`; We can get this information ourselves using the stop_id and timestamp fields for each MDS trip_id so this field doesn't need to be part of the spec

**MDS Telemetry**
- remove `gtfs_route_id`, `gtfs_trip_id`, `gtfs_service_id`; These can all be found by linking to MDS trips with MDS trip_id.

**MDS Trips**
- remove `gtfs_route_id`, `gtfs_service_id`, `gtfs_shape_id`; These can all be found by linking to GTFS trips with GTFS trip_id and the GTFS API URL.

The remaining issue is `gtfs_stop_id` in **MDS Events** and **MDS Telemetry**. Since SFMTA is planning to use MDS stop_id here, we don't need a gtfs_stop_id field. I won't add it back to the pull request for now, but if you want to add it back as an optional field that other jurisdictions can use that works. Let me know and I will update the pull request to have it as optional.

schnuerle commented:
Thank you looks great and is much simpler! 

For `gtfs_stop_id` in **MDS Events** and **MDS Telemetry**, I feel like that is what SFMTA (and others) should use, instead of MDS' `stop_id`. 

MDS `stop_id` is a UUID, and explicitly references the [Stops object](https://github.com/openmobilityfoundation/mobility-data-specification/blob/dev/data-types.md#stops), so it would not work as the stop identifier in GTFS that you would be interested in knowing about for Events and Telemetry.

matthewmarwedel commented:
@schnuerle The plan is for SFMTA to actually put all our stops into MDS Stops endpoint and use those, but I will add back in `gtfs_stop_id` to **Events, Telemetry** as an optional field in case another jurisdiction wants to use it or SFMTA decides to use it in the future.

schnuerle commented:
Thank you Matt, that makes sense to both have it there to connect to GTFS data, or use MDS Stops.

How are you planning on creating each stop for each route for each operator using MDS Stops? Are all the data fields you need for that in MDS Stops already? Anything else you'd need to support this?

schnuerle commented:
Ok that makes sense for your use of MDS Stops. And I think it's good to add back gtfs_stop_id.

I am not 100% on the reasoning why MDS `stop_id` is in Telemetry, but not Events. I think it should be added as optional to Events too. You or I can add that as part of this pull request, and look for feedback from the working group on this.

matthewmarwedel commented:
Hi @schnuerle I wanted to update you that I am planning to leave SFMTA at the end of this week and go to a new position elsewhere. If it's possible to get this pull request merged by the end of the week that would be great. I do feel that it's been thoroughly reviewed and it's ready for merge with the dev branch. If not, the Github handles for the rest of the SFMTA team are:  @Mu-yi-Zhou, @monachiu, and @alexdemisch. Please contact them for additional questions on the pull request.

schnuerle commented:
Hi @matthewmarwedel yes, agree it can be pushed to dev. We had a recap at the [last working group meeting](https://github.com/openmobilityfoundation/mobility-data-specification/wiki/Web-conference-notes,-2025.03.13-(MDS-Working-Group)) (see link for [slide](https://docs.google.com/presentation/d/1QZRGnWlswlQOphkAKKFZidIxueb8zLJgGwozmUR_B3E/edit?slide=id.g333f736bb8c_0_20#slide=id.g333f736bb8c_0_20) and recording) about the status of it and gave people another full week to comment. So I will review and push to dev. Thanks for your work and good luck in your new endeavor!

ericyco commented:
Hi @schnuerle, one issue that's come up is that MDS gtfs_trip_id requires a UUID while GTFS doesn't.

schnuerle commented:
@ericyco You are absolutely right, I can't believe I didn't see that sooner. It should be a string to keep it generic enough. I'll update.

schnuerle commented:
@ericyco take a look and check my work if you can, for the updates here at PR #937.

ericyco commented:
@schnuerle looks great, thanks for the quick update on both gtfs_trip_id and gtfs_stop_id!

schnuerle commented:
Alright, merged to dev now, thanks for the catch!
